File: leetcode/reorder-list/reorder-list.cpp
/**************************************
* @author dujiong
* @date 2016.9.20
* @version V0.1
**************************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode *head) {
        if(head == NULL || head->next == NULL){
            return;
        }
        ListNode *p = head;
        ListNode *q = head->next;
        while(q && q->next){
            p = p->next;
            q = q->next->next;
        }
        ListNode *right = p->next;
       	p->next = NULL;
        ListNode *res = head;
        right = reverse(right);
        while(right!=NULL){
            ListNode* temp = res->next;
            res->next = right;
            ListNode* tempRight = right->next;
            right->next = temp;
            res = res->next->next;
            right = tempRight;
        }
    }
private:
    ListNode* reverse(ListNode* head){
        if(head==NULL){
            return NULL;
        }
        ListNode* pNext = NULL;
        ListNode* pPrev = NULL;
        while(head!=NULL){
            pNext = head->next;
            head->next = pPrev;
            pPrev = head;
            head = pNext;
        }
        return pPrev;
    }
};	

File: leetcode-golang/Reorder List/Reorder List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
//1、将原链表分成两部分
//2、第二部分链表反转
//3、将第二部分链表间隔地插入第一部分链表中
func reorderList(head *ListNode) {
	if head == nil || head.Next == nil || head.Next.Next == nil {
		return
	}
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	rightHead := slow.Next
	slow.Next = nil
	//第二部分反转之后的链表
	reverseRightHead := reverseList(rightHead)
	leftHead := head
	for reverseRightHead != nil {
		leftNext, rightNext := leftHead.Next, reverseRightHead.Next
		leftHead.Next = reverseRightHead
		reverseRightHead.Next = leftNext
		leftHead = leftNext
		reverseRightHead = rightNext
	}
}

func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	var prev *ListNode
	for head != nil {
		next := head.Next
		head.Next = prev
		prev = head
		head = next
	}
	return prev
}